p2p/protocol/identify/id.go

```diff
@@ -146,40 +154,96 @@ func NewIDService(h host.Host, opts ...Option) *IDService {
 
 	// register protocols that do not depend on peer records.
 	h.SetStreamHandler(IDDelta, s.deltaHandler)
-	h.SetStreamHandler(LegacyID, s.requestHandler)
-	h.SetStreamHandler(LegacyIDPush, s.pushHandler)
-
-	// register protocols that depend on peer records.
-	h.SetStreamHandler(ID, s.requestHandler)
+	h.SetStreamHandler(ID, s.sendIdentifyResp)
 	h.SetStreamHandler(IDPush, s.pushHandler)
 
 	h.Network().Notify((*netNotifiee)(s))
 	return s
 }
 
-func (ids *IDService) handleEvents() {
+func (ids *IDService) loop() {
 	defer ids.refCount.Done()
 
+	phs := make(map[peer.ID]*peerHandler)
 	sub, err := ids.Host.EventBus().Subscribe([]interface{}{&event.EvtLocalProtocolsUpdated{},
 		&event.EvtLocalAddressesUpdated{}}, eventbus.BufSize(256))
 	if err != nil {
 		log.Errorf("failed to subscribe to events on the bus, err=%s", err)
 		return
 	}
 
-	defer sub.Close()
+	defer func() {
+		sub.Close()
+		for pid := range phs {
+			phs[pid].close()
+		}
+	}()
+
+	phClosedCh := make(chan peer.ID)
 
 	for {
 		select {
+		case addReq := <-ids.addPeerHandlerCh:
+			rp := addReq.s.Conn().RemotePeer()
+			ph, ok := phs[rp]
+			if ok {
+				addReq.resp <- ph
+				continue
+			}
+
+			if ids.Host.Network().Connectedness(rp) == network.Connected {
+				mes := &pb.Identify{}
+				ids.populateMessage(mes, addReq.s.Conn())
+				ph = newPeerHandler(rp, ids, mes)
+				phs[rp] = ph
+				addReq.resp <- ph
+			}
+
+		case rmReq := <-ids.rmPeerHandlerCh:
+			rp := rmReq.p
+			if ids.Host.Network().Connectedness(rp) != network.Connected {
+				// before we remove the peerhandler, we should ensure that it will not send any
+				// more messages. Otherwise, we might create a new handler and the Identify response
+				// synchronized with the new handler might be overwritten by a message sent by this "old" handler.
+				ids.refCount.Add(1)
+				go func(req *rmPeerHandlerReq, ph *peerHandler) {
+					defer ids.refCount.Done()
+					if ph != nil {
+						ph.close()
+						select {
+						case <-ids.ctx.Done():
+							return
+						case phClosedCh <- req.p:
+						}
+					}
+				}(rmReq, phs[rp])
+			}
+
+		case p := <-phClosedCh:
+			delete(phs, p)
```

Comment:
I think we need to check if we're connected to the peer at this point. If we are, we need to restart this peer handler.

Reply:
@Stebalien Great find ! This is done.

Reply:
If we do that, we may need to also "forward" any pending push/delta signals.

On the other hand, we could restart it by:

1. Not starting inside `newPeerHandler`, adding a `start` method instead.
2. Having the `start` function initialize the ctx + cancel fields and start the loop.

But that's not necessary.

Reply:
@Stebalien After thinking through this, I think your approach of having a `start` function that we can use to re-start the handler is better as it means we don't have to take a read lock inside the event loop to copy the state. 

I have made the change.